components/canvas/webgl_paint_thread.rs

```diff
@@ -122,47 +114,61 @@ impl WebGLPaintThread {
     }
 
+    fn read_back_pixels(&self) -> Vec<u8> {
+        let width = self.size.width as usize;
+        let height = self.size.height as usize;
+
+        let mut pixels = gl::read_pixels(0, 0,
+                                         self.size.width as gl::GLsizei,
+                                         self.size.height as gl::GLsizei,
+                                         gl::RGBA, gl::UNSIGNED_BYTE);
+        // flip image vertically (texture is upside down)
+        let orig_pixels = pixels.clone();
+        let stride = width * 4;
+        for y in 0..height {
+            let dst_start = y * stride;
+            let src_start = (height - y - 1) * stride;
+            let src_slice = &orig_pixels[src_start .. src_start + stride];
+            (&mut pixels[dst_start .. dst_start + stride]).clone_from_slice(&src_slice[..stride]);
+        }
+
+        // rgba -> bgra
+        byte_swap(&mut pixels);
+
+        pixels
+    }
+
     fn send_data(&mut self, chan: IpcSender<CanvasData>) {
         match self.data {
-            WebGLPaintTaskData::Readback(_, ref webrender_api_and_image_key) => {
-                let width = self.size.width as usize;
-                let height = self.size.height as usize;
-
-                let mut pixels = gl::read_pixels(0, 0,
-                                                 self.size.width as gl::GLsizei,
-                                                 self.size.height as gl::GLsizei,
-                                                 gl::RGBA, gl::UNSIGNED_BYTE);
-                // flip image vertically (texture is upside down)
-                let orig_pixels = pixels.clone();
-                let stride = width * 4;
-                for y in 0..height {
-                    let dst_start = y * stride;
-                    let src_start = (height - y - 1) * stride;
-                    let src_slice = &orig_pixels[src_start .. src_start + stride];
-                    (&mut pixels[dst_start .. dst_start + stride]).clone_from_slice(&src_slice[..stride]);
-                }
+            WebGLPaintTaskData::Readback(..) => {
+                let pixels = self.read_back_pixels();
 
-                // rgba -> bgra
-                byte_swap(&mut pixels);
+                let pixel_data = CanvasPixelData {
+                    image_data: IpcSharedMemory::from_bytes(&pixels[..]),
+                    image_key: None,
+                };
 
-                if let Some((ref wr, wr_image_key)) = *webrender_api_and_image_key {
-                    // TODO: This shouldn't be a common path, but try to avoid
-                    // the spurious clone().
-                    wr.update_image(wr_image_key,
-                                    width as u32,
-                                    height as u32,
+                chan.send(CanvasData::Pixels(pixel_data)).unwrap();
+            }
+            WebGLPaintTaskData::WebRender(ref wr_api, _, Some(image_key)) => {
```

Comment:
Hm... Yes, though you'd need to reach the actual context, plus apply commands locally in this thread instead of sending them to WebRender.

All in all, I think the best approach is to expose the RendererKind (and probably the wrapper types for the context too) and move the fallback logic here, so that the `Readback` data contains the correct kind of context.